qiskit_experiments/analysis/curve_fit_analysis.py

```diff
@@ -0,0 +1,196 @@
+            func: fit function `f(x, *params)`.
+            p0: Optional, initial parameter values for curve_fit.
+            p0_func: Optional, function for calculating initial p0.
+            bounds: Optional, parameter bounds for curve_fit.
+            fit_mean_data: Optional, if True pass means of data points to curve_fit.
+            plot: If True generate a plot of fitted data.
+            ax: Optional, matplotlib axis to add plot to.
+
+        Returns:
+            tuple: A pair ``(analysis_results, figures)`` where
+                   ``analysis_results`` may be a single or list of
+                   AnalysisResult objects, and ``figures`` may be
+                   None, a single figure, or a list of figures.
+        """
+        # Compute curve fit data
+        xdata, ydata, ydata_sigma = self._curve_fit_data(experiment_data.data)
+        if fit_mean_data:
+            xvals, yvals, sigma = self._mean_data(xdata, ydata)
+        else:
+            xvals, yvals, sigma = xdata, ydata, ydata_sigma
+
+        if p0_func is not None and p0 is None:
+            p0 = p0_func(xdata, ydata, sigma=sigma)
+
+        # Run curve fit
+        popt, pcov = curve_fit(func, xvals, yvals, sigma=sigma, p0=p0, bounds=bounds)
+        popt_err = np.sqrt(np.diag(pcov))
+        chisq = self._chisq(xvals, yvals, sigma, func, popt)
+
+        result = AnalysisResult(
+            {
+                "popt": popt,
+                "popt_err": popt_err,
+                "pcov": pcov,
+                "chisq": chisq,
+            }
+        )
+
+        if plot and HAS_MATPLOTLIB:
+            mean_data = (xvals, yvals, sigma) if fit_mean_data else None
+            ax = self._curve_fit_plot(func, popt, popt_err, xdata,
+                                      ydata=ydata, mean_data=mean_data, ax=ax)
+            # TODO: figure out what to do with plots
+            return result, [ax]
+
+        return result, None
+
+    @staticmethod
+    def _counts_probability(counts: "Counts", key: Union[str, int]) -> Tuple[float]:
+        """Return the specified outcome probability mean and variance"""
+        shots = sum(counts.values())
+        p_mean = counts.get(key, 0.0) / shots
+        p_var = shots * p_mean * (1 - p_mean)
+        return p_mean, p_var
+
+    @staticmethod
+    def _chisq(
+        xdata: np.ndarray, ydata: np.ndarray, sigma: np.ndarray, func: Callable, popt: np.ndarray
+    ) -> float:
+        """Return the chi-squared of fit"""
+        yfits = func(xdata, *popt)
+        residuals = ((yfits - ydata) / sigma) ** 2
+        return np.sum(residuals)
+
+    @classmethod
+    def _curve_fit_data(cls, data: List[Dict[str, any]]) -> Tuple[np.ndarray]:
+        """Return input data for curve_fit function.
+
+        This requires count data and metadata with `"xdata"`, `"ylabel"`
+        keys containing the numeric x-value for fitting, and the outcome
+        bitstring for computing y-value probability from counts.
```

Comment:
What if you are not dealing with counts data? E.g. IQ data or something derived from IQ data?

Reply:
This is WIP. It only deals with count data at the moment, but intent is you would have functions that compute probs from IQ data too.

Reply:
Have a look at #20 it separates the data processing from the fitting.